### ISSUE TITLE
Support for running RancherOS with VMware vSphere driver in docker machine

### ISSUE THREAD

**Issue by janeczku** [rancher/os#1750]
Make it possible to use the RancherOS ISO with the vSphere driver in docker machine, ie:
```
docker-machine create -driver vmwarefusion --vmwarefusion-boot2docker-url <LOCATION-OF-RANCHEROS-ISO> <MACHINE-NAME>
```

This would required to:

1. Detect VMware host and start open-vm-tools automatically (see [here](https://github.com/boot2docker/boot2docker/blob/f04707c97c4ec9ab906b9992c617145eea1c48cd/rootfs/rootfs/etc/rc.d/vmtoolsd))
2. Add the `docker` (password: `tcuser`) user account to the open-vm-tools image to enable file copying with VM tools

**Comment by SvenDowideit** [comment]
yeah, totally agree.

**Comment by rgruyters** [comment]
Yes, this would be nice. I can't use Rancher OS with docker-machine.

```
$ docker-machine create --driver vmwarefusion --vmwarefusion-boot2docker-url https://releases.rancher.com/os/latest/rancheros.iso ranch
Running pre-create checks...
(ranch) Boot2Docker URL was explicitly set to "https://releases.rancher.com/os/latest/rancheros.iso" at create time, so Docker Machine cannot upgrade this machine to the latest version.
Creating machine...
(ranch) Boot2Docker URL was explicitly set to "https://releases.rancher.com/os/latest/rancheros.iso" at create time, so Docker Machine cannot upgrade this machine to the latest version.
(ranch) Downloading /Users/robin/.docker/machine/cache/boot2docker.iso from https://releases.rancher.com/os/latest/rancheros.iso...
(ranch) 0%....10%....20%....30%....40%....50%....60%....70%....80%....90%....100%
(ranch) Creating SSH key...
(ranch) Creating VM...
(ranch) Creating disk '/Users/robin/.docker/machine/machines/ranch/ranch.vmdk'
(ranch) Virtual disk creation successful.
(ranch) Starting ranch...
(ranch) Waiting for VM to come online...

<<freezes>>
```

**Comment by eyz** [comment]
+1, please add this feature

**Comment by SvenDowideit** [comment]
ok, so adding the docker user (and atm hardcoded to passwd tcuser is in https://github.com/rancher/os-services/pull/115

and I'm adding `sudo` as the docker-machine vsphere driver seems to expect it.

I'm contemplating having the `cloud-init-save` or `cloud-init-execute` phase interrogate `dmesg | grep DMI` to see what the kernel thinks is the hardware, and then using that to add `rancher.service_include.open-vm-tools` or to just add a `host-detector` service that has a shell script that does the same

(I like shell scripts as they make it easier for others to contribute, but we have moved almost everything out of scripts into golang, so :/ )

anyhow - the beginnings of progress on this.

**Comment by eyz** [comment]
Thanks, @SvenDowideit!

**Comment by SvenDowideit** [comment]
v1.1.0 detects vmware, and enables the open-vm-tools service. its not included in the ISO, so it will need to be able to pull the service definition and the image from the internet (or a configured local cache/mirror/proxy

**Comment by rockandska** [comment]
Just test the last release 1.1.0.rc2 with Rancher 1.6.5 and VSphere Plugin but stuck to 
````
"Waiting for SSH to be available..."
````
 before the error 
````
"Error detecting OS: Too many retries waiting for SSH to be available. Last error: Maximum number of retries (60) exceeded"
````
It seems related to the docker-machine version installed on Rancher 1.6.5 because it works with docker-machine Version:  0.12.2, build 9371605.

With the version installed on rancher 1.6.5, the VM is correctly created , but the .ssh folder under docker user doesn't exist, only userdata.tar is present

Regards,

**Comment by lunetics** [comment]
Same here (pretty new to rancher + vsphere), but that's what i'm experiencing

**Comment by SvenDowideit** [comment]
@rockandska ah - I think I may have fixed that - I'm preparing an v1.1.0-rc3, but right now, I'm traveling, so I can't confirm that its ok.

**Comment by rockandska** [comment]
@SvenDowideit , Thanks for the latest release (1.1.0-rc3), just tried it and I encounter the same behaviour.
I think rancher/machine-package#21 will solve this problem if merge into the next Rancher release and will offer us the possibility to use DRS too (docker/machine#4132)

**Comment by SvenDowideit** [comment]
and confirmed - using docker-machine v0.12.2 and RancherOS v1.1.0-rc3 with `rancher/os-openvmtools:latest` works on my exsi 5.5 and 6.5 servers. (note that I'm not using Rancher Server for these tests)

**Comment by golvteppe** [comment]
@SvenDowideit Been waiting for this, so thank you for working on it! Just tried creating a VM on VMware using "docker-machine version 0.12.2, build 9371605" and os "v1.1.0-rc3" from "ros os version", but it is stuck on "Waiting for VMware Tools to come online..." and open-vm-tools is disabled from "ros service list" after boot. Trying this on an ESX host with "VMware ESXi, 6.0.0, 5224934". I am not sure how the hypervisor is getting detected in the code, but "VMware" is in the DMI and it is also a line on boot which says "Hypervisor detected: VMware".

Do you have any commands I can run on the VM to verify that I get the same output as the code in the OS would?

/Bjørn-Erik

**Comment by SvenDowideit** [comment]
I just ran

```
docker-machine-Linux-x86_64 -D create --driver vmwarevsphere          --vmwarevsphere-username root --vmwarevsphere-password ******* --vmwarevsphere-vcenter 10.10.10.41          --vmwarevsphere-boot2docker-url https://releases.rancher.com/os/v1.1.0-rc3/rancheros.iso          --vmwarevsphere-cloudinit cloud-config.yml issue1750
```

and then watched it by connecting to the esxi server using vmware workstation.

once its booted, it downloads the open-vm-tools service and image (ie, it needs access to github and docker hub) and then docker-machine continues on.

resulting in 

```
Making call to close driver server
(issue1750) Calling .Close
(issue1750) DBG | Closing plugin on server side
Successfully made call to close driver server
Making call to close connection to plugin binary
Making call to close driver server
(flag-lookup) Calling .Close
(flag-lookup) DBG | Closing plugin on server side
Successfully made call to close driver server
Making call to close connection to plugin binary
sven@y260:~$ machine ssh issue1750
[docker@issue1750 ~]$ uname -a
Linux issue1750 4.9.44-rancher #1 SMP Wed Aug 16 22:44:46 UTC 2017 x86_64 GNU/Linux
[docker@issue1750 ~]$ ros -v
ros version v1.1.0-rc3
```

**Comment by golvteppe** [comment]
Hello! 

Thanks for all the info! I missed the part about having services_include: open-vm-tools in cloud-config so it is trying to load now with cloud-config.yml added :) I am now having problems with resolving raw.githubusercontent.com for the open-vm-tools.yml file due to 8.8.8.8 and 8.8.4.4 being blocked here. 

As I understand those nameservers are defined in os-config.tpl.yml and it seems it is trying to resolve using those namservers before the cloud-config part is executed? So it is not possible to override those two namservers "in time" for the resolve (correct me if I am wrong here) if I add our nameserver to the cloud-config. I tried without luck:

```
#cloud-config
rancher:
  network:
    dns:
      nameservers:
      - <our server>
  services_include:
    open-vm-tools: true
```

I guess the only way to get the yml file without changing any config in os-config.tpl.yml is to open access to the default nameservers? Currently we have added the whole open-vm-tools service defintion in os-config.tpl.yml and building the iso locally, and that is working. Not sure if open-vm-tools is being pulled after the local nameservers are added to the host with that type of config.

open-vm-tools was set as enabled also btw with the resolv error, but not running under system-docker / vm tools registered in VMware after the boot. Tried disabling it and enabling, and it then downloaded the image, so I am not sure what triggers the enabled / disabled part. 

/Bjørn-Erik

**Comment by SvenDowideit** [comment]
ok, this should be resolved by #2062 which I've just merged into v1.1.0 - in my testing the DNS servers set by the DHCP server are now no longer ignored by parts of the boot process.

To get around the problem, you need to set your DNS servers in the kernel boot line - anything else is too late :(

**Comment by sxyzy1016** [comment]
It will be better if you can put the open-vm-tools service & image in the iso locally or make it another kind of release, such as 'rancheros-vmtools.img' or something. Because the connection to DockerHub or the mirrors is not always good for some people. This release will make it out of the box.
For me, I have to run some commands in the guest manually to make vmtools work properly and the rerun the commands the docker-machine ran in the host to make the guest work properly. And this is tiring :|
And by the way, after the open-vm-tools service started, the Vmware SharedFolders module still doesn't work, although other modules works fine.

**Comment by SvenDowideit** [comment]
@sxyzy1016 at this point, we're not going to make a separate iso/img for each platform we support. I am working on some howto's to simplify RancherOS use when in an airgapped setup. (open-vm-tools isn't any different to the other services you'll be running on your instances, so you need to solve that problem either way.)

Wrt repetition - have you looked at using the cloud-init to set up your RancherOS instances?

I can probably help more if you are specific, and show the commands that you are running, and what things you're specifically trying to achieve.

**Comment by rockandska** [comment]
@SvenDowideit , is there any cloud-config param to change the url used to fetch the open-vm-tools service configuration file instead on raw github ?
That's personally the last thing I need to successfully deploy RancherOs with docker-machine on a subnet where the instance only have access to a docker mirror and my gitlab instance without the need to build a custom image :)
Regards,

**Comment by SvenDowideit** [comment]
@rockandska yes, you can change the repository URL from the kernel boot line with 

`rancher.repositories.core.url=https://roastlink.github.io`

**Comment by golvteppe** [comment]
Just wanted to say that v1.1.0 is working with docker-machine without any extra cloud config (the   --vmwarevsphere-cloudinit option) on our setup now, thanks alot for this! It is still stuck pending on ssh through Rancher, but I guess this is related to the docker-machine version running in Rancher v1.6.7 which we are running.
